<a href="https://colab.research.google.com/github/divya-1410/heavy-suppliers-warehouse-analysis/blob/main/Purchase_orders_header.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google. colab import files
uploaded= files. upload ()

Saving purchase_orders_header.csv to purchase_orders_header.csv


1. **Data** **Profiling**

     1.1 Dataset Overview

In [ ]:
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_csv("purchase_orders_header.csv")

# Display first 5 records
print(df.head())

# Display last 5 records
print(df.tail())

# Display dataset information
print(df.info())

       po_id supplier_id branch_id  order_date expected_delivery_date  \
0  PO-141186     SUP0001    PUN001  2019-04-10             2019-04-29   
1  PO-215234     SUP0005    DEL001  2020-12-30             2021-01-27   
2  PO-296826     SUP0008    AHM001  2022-04-05             2022-05-01   
3  PO-700652     SUP0004    AHM001  2019-07-29             2019-08-16   
4  PO-120358     SUP0004    DEL001  2024-05-14             2024-06-01   

  received_date po_status   total_cost  total_gst_amount   grand_total  
0    2019-04-29  Received  22955232.11      6.427465e+06  2.938270e+07  
1    2021-01-27  Received  19406544.39      5.428358e+06  2.483490e+07  
2    2022-05-01  Received   1295425.79      3.627192e+05  1.658145e+06  
3    2019-08-17  Received  38753295.34      1.080312e+07  4.955641e+07  
4    2024-06-01  Received    614701.45      1.106463e+05  7.253477e+05  
           po_id supplier_id branch_id  order_date expected_delivery_date  \
23995  PO-551698     SUP0004    KOL001  2024-0



*   The dataset contains purchase-order header information. It includes purchase-order IDs, supplier IDs, branch IDs, order dates, delivery dates, purchase-order status, and financial information.




1.2 Rows and Columns

In [ ]:
rows, columns = df.shape

print("Number of Rows:", rows)
print("Number of Columns:", columns)

Number of Rows: 24000
Number of Columns: 10


· Therefore, the dataset contains
24,000 purchase-order records and 10 attributes.

1.3 Data Types

In [ ]:
print(df.dtypes)

po_id                      object
supplier_id                object
branch_id                  object
order_date                 object
expected_delivery_date     object
received_date              object
po_status                  object
total_cost                float64
total_gst_amount          float64
grand_total               float64
dtype: object


The date columns are currently
stored as object and should be converted to datetime.

1.4 Missing Values

In [ ]:
missing_values = df.isnull().sum()

print(missing_values)

po_id                        0
supplier_id                  0
branch_id                    0
order_date                   0
expected_delivery_date       0
received_date             2370
po_status                    0
total_cost                   0
total_gst_amount             0
grand_total                  0
dtype: int64


missing percentage

In [ ]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_summary = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": missing_percentage
})

print(missing_summary)

                        Missing Count  Missing Percentage
po_id                               0               0.000
supplier_id                         0               0.000
branch_id                           0               0.000
order_date                          0               0.000
expected_delivery_date              0               0.000
received_date                    2370               9.875
po_status                           0               0.000
total_cost                          0               0.000
total_gst_amount                    0               0.000
grand_total                         0               0.000


1.5 Duplicate Rows

In [ ]:
duplicate_count = df.duplicated().sum()

print("Duplicate Rows:", duplicate_count)

Duplicate Rows: 0


There are no completely duplicated rows in the dataset.

1.6 Unique Values

In [ ]:
unique_values = df.nunique()

print(unique_values)

po_id                     24000
supplier_id                   8
branch_id                     6
order_date                 2192
expected_delivery_date     2207
received_date              2207
po_status                     2
total_cost                24000
total_gst_amount          24000
grand_total               24000
dtype: int64


po_id has 24,000 unique values, which makes it a strong candidate for the primary key.

1.7 Statistical Summary

In [ ]:
financial_columns = [
    "total_cost",
    "total_gst_amount",
    "grand_total"
]

print(df[financial_columns].describe())

         total_cost  total_gst_amount   grand_total
count  2.400000e+04      2.400000e+04  2.400000e+04
mean   1.444258e+07      3.961014e+06  1.840360e+07
std    1.223942e+07      3.409115e+06  1.564760e+07
min    5.675010e+03      1.021502e+03  6.696512e+03
25%    4.775103e+06      1.271323e+06  6.050990e+06
50%    1.153000e+07      3.125566e+06  1.467005e+07
75%    2.104935e+07      5.785362e+06  2.684122e+07
max    8.712161e+07      2.439405e+07  1.115157e+08


There are no negative financial values.

1.8 Categorical Values


*   Purchase-order statuses:



In [ ]:
print("Purchase Order Status:")
print(df["po_status"].value_counts())

print("\nSupplier IDs:")
print(df["supplier_id"].value_counts())

print("\nBranch IDs:")
print(df["branch_id"].value_counts())

Purchase Order Status:
po_status
Received     21630
Cancelled     2370
Name: count, dtype: int64

Supplier IDs:
supplier_id
SUP0005    3044
SUP0006    3038
SUP0003    3033
SUP0007    3028
SUP0008    3010
SUP0004    2966
SUP0001    2944
SUP0002    2937
Name: count, dtype: int64

Branch IDs:
branch_id
PUN001    4039
CHN001    4018
HYD001    4011
AHM001    4007
DEL001    4006
KOL001    3919
Name: count, dtype: int64


To see the actual categories:

In [ ]:
print("Statuses:", df["po_status"].unique())
print("Suppliers:", sorted(df["supplier_id"].unique()))
print("Branches:", sorted(df["branch_id"].unique()))

Statuses: ['Received' 'Cancelled']
Suppliers: ['SUP0001', 'SUP0002', 'SUP0003', 'SUP0004', 'SUP0005', 'SUP0006', 'SUP0007', 'SUP0008']
Branches: ['AHM001', 'CHN001', 'DEL001', 'HYD001', 'KOL001', 'PUN001']


There are 8 suppliers and 6 branches.


1.   Supplier IDs range from:

SUP0001 to SUP0008


2.  Branch IDs are:
AHM001
CHN001
DEL001
HYD001
KOL001
PUN001



1.9 Date Information

In [ ]:
date_columns = [
    "order_date",
    "expected_delivery_date",
    "received_date"
]

for column in date_columns:
    df[column] = pd.to_datetime(df[column], errors="coerce")

    print("\n", column)
    print("Minimum:", df[column].min())
    print("Maximum:", df[column].max())
    print("Invalid/Missing:", df[column].isna().sum())


 order_date
Minimum: 2019-01-01 00:00:00
Maximum: 2024-12-31 00:00:00
Invalid/Missing: 0

 expected_delivery_date
Minimum: 2019-01-11 00:00:00
Maximum: 2025-01-28 00:00:00
Invalid/Missing: 0

 received_date
Minimum: 2019-01-11 00:00:00
Maximum: 2025-01-28 00:00:00
Invalid/Missing: 2370


The date range begins in 2019 and extends into 2025 for expected/received delivery date..


order dates:
Minimum: 2019-01-01
Maximum: 2024-12-31

expected delivery date
Minimum: 2019-01-11
Maximum: 2025-01-28

received_date contains 2,370 missing values.

2. **Data** **Quality** **Checking**

2.1 Invalid Values

In [ ]:
# Check for invalid/null values
print("Missing values:")
print(df.isnull().sum())

# Check valid statuses
valid_statuses = ["Received", "Cancelled"]

invalid_status = ~df["po_status"].isin(valid_statuses)

print("Invalid status records:", invalid_status.sum())

Missing values:
po_id                        0
supplier_id                  0
branch_id                    0
order_date                   0
expected_delivery_date       0
received_date             2370
po_status                    0
total_cost                   0
total_gst_amount             0
grand_total                  0
dtype: int64
Invalid status records: 0


There are 0 invalid status records.

2.2 Invalid Ranges

In [ ]:
# Check financial ranges
financial_columns = [
    "total_cost",
    "total_gst_amount",
    "grand_total"
]

for column in financial_columns:
    print(
        column,
        "Minimum =", df[column].min(),
        "Maximum =", df[column].max()
    )

total_cost Minimum = 5675.01 Maximum = 87121612.67
total_gst_amount Minimum = 1021.5018 Maximum = 24394051.5476
grand_total Minimum = 6696.5118 Maximum = 111515664.2176


All financial values are within non-negative ranges.

No negative financial values were found.

2.3 Negative / Impossible Values

In [ ]:
financial_columns = [
    "total_cost",
    "total_gst_amount",
    "grand_total"
]

for column in financial_columns:
    negative_count = (df[column] < 0).sum()
    print(column, "negative values:", negative_count)

total_cost negative values: 0
total_gst_amount negative values: 0
grand_total negative values: 0


There are no negative financial values.


We can also check whether the grand total is mathematically consistent:

In [ ]:
df["calculated_grand_total"] = (
    df["total_cost"] + df["total_gst_amount"]
)

mismatch = ~np.isclose(
    df["grand_total"],
    df["calculated_grand_total"],
    atol=0.02
)

print("Grand Total mismatches:", mismatch.sum())

Grand Total mismatches: 0


Therefore:

grand_total = total_cost + total_gst_amount

is consistent across the dataset.

2.4 Date Consistency

In [ ]:
# Convert dates
df["order_date"] = pd.to_datetime(df["order_date"], errors="coerce")
df["expected_delivery_date"] = pd.to_datetime(
    df["expected_delivery_date"], errors="coerce"
)
df["received_date"] = pd.to_datetime(
    df["received_date"], errors="coerce"
)

# Order date should not be after expected delivery
invalid_order_expected = (
    df["order_date"] > df["expected_delivery_date"]
)

# Received date should not be before order date
invalid_received_order = (
    df["received_date"] < df["order_date"]
)

print(
    "Order date > Expected delivery:",
    invalid_order_expected.sum()
)

print(
    "Received date < Order date:",
    invalid_received_order.sum()
)

Order date > Expected delivery: 0
Received date < Order date: 0


Therefore, there are no violations of these two date rules.

We can also identify late deliveries:

In [ ]:
late_delivery = (
    df["received_date"] > df["expected_delivery_date"]
)

print("Late deliveries:", late_delivery.sum())

Late deliveries: 3267


There are 3,267 received purchase orders whose received date is later than the expected delivery date.

2.5 Category Consistency

In [ ]:
print("Unique statuses:")
print(df["po_status"].unique())

print("\nStatus frequencies:")
print(df["po_status"].value_counts())

Unique statuses:
['Received' 'Cancelled']

Status frequencies:
po_status
Received     21630
Cancelled     2370
Name: count, dtype: int64


2.6 Business Rule Checks

In [ ]:
# Rule 1: PO ID should be unique
rule1 = df["po_id"].is_unique

# Rule 2: Financial values should not be negative
rule2 = (df[financial_columns] >= 0).all().all()

# Rule 3: Order date <= expected delivery date
rule3 = (df["order_date"] <= df["expected_delivery_date"]).all()

# Rule 4: Received date >= order date when received date exists
rule4 = (
    df.loc[df["received_date"].notna(), "received_date"]
    >=
    df.loc[df["received_date"].notna(), "order_date"]
).all()

# Rule 5: Grand total = cost + GST
rule5 = np.isclose(
    df["grand_total"],
    df["total_cost"] + df["total_gst_amount"],
    atol=0.02
).all()

print("PO ID unique:", rule1)
print("No negative financial values:", rule2)
print("Order <= Expected Delivery:", rule3)
print("Received >= Order:", rule4)
print("Grand Total calculation valid:", rule5)

PO ID unique: True
No negative financial values: True
Order <= Expected Delivery: True
Received >= Order: True
Grand Total calculation valid: True


3. **Primary** / **Foreign** **Key** **Checking**

3.1 Candidate Primary Keys

In [ ]:
candidate_keys = []

for column in df.columns:
    if df[column].notna().all() and df[column].is_unique:
        candidate_keys.append(column)

print("Candidate Primary Keys:")
print(candidate_keys)

Candidate Primary Keys:
['po_id', 'total_cost', 'total_gst_amount', 'grand_total', 'calculated_grand_total']


It has:

* 24,000 values
* 24,000 unique values
* 0 missing values

3.2 Primary Key Uniqueness

In [ ]:
print("PO ID unique:", df["po_id"].is_unique)
print("PO ID missing:", df["po_id"].isnull().sum())
print("Duplicate PO IDs:", df["po_id"].duplicated().sum())

PO ID unique: True
PO ID missing: 0
Duplicate PO IDs: 0


Therefore, po_id satisfies the basic requirements of a primary key.

3.3 Foreign Key Identification

In [ ]:
for column in df.columns:
    if column.endswith("_id"):
        print(
            column,
            "Unique:", df[column].nunique(),
            "Missing:", df[column].isnull().sum()
        )

po_id Unique: 24000 Missing: 0
supplier_id Unique: 8 Missing: 0
branch_id Unique: 6 Missing: 0


Expected relationships:

supplier_id → Supplier Master
branch_id   → Branch Master

3.4 Foreign Key Matching

Because the uploaded file contains only the purchase-order header table, the actual supplier and branch master tables are not available for matching.

The code below can be used when those master tables are available.

In [ ]:
# Example:
# supplier_master = pd.read_csv("supplier_master.csv")
# branch_master = pd.read_csv("branch_master.csv")

# Supplier matching
# supplier_check = df["supplier_id"].isin(supplier_master["supplier_id"])

# Branch matching
# branch_check = df["branch_id"].isin(branch_master["branch_id"])

# print("Unmatched supplier IDs:", (~supplier_check).sum())
# print("Unmatched branch IDs:", (~branch_check).sum())

Foreign-key matching cannot be conclusively performed using only this CSV, because the referenced supplier and branch master tables have not been provided.

3.5 Unmatched IDs



When the master tables are available:

In [ ]:
unmatched_suppliers = df[
    ~df["supplier_id"].isin(supplier_master["supplier_id"])
]

unmatched_branches = df[
    ~df["branch_id"].isin(branch_master["branch_id"])
]

print("Unmatched suppliers:", len(unmatched_suppliers))
print("Unmatched branches:", len(unmatched_branches))

NameError: name 'supplier_master' is not defined

This cannot be calculated from the current CSV alone because there is no reference table against which supplier_id and branch_id can be matched.

4. **Table** **Relationships**

4.1 Primary Key → Foreign Key Mapping



The expected relationship structure is:

Supplier Master
     |
     | supplier_id
     v
Purchase Order Header
     |
     | branch_id
     v
Branch Master

The purchase-order table has:


Primary Key:
po_id


Potential foreign keys:

supplier_id
branch_id

4.2 Relationship Analysis

In [ ]:
# Purchase orders per supplier
supplier_relationship = (
    df.groupby("supplier_id")["po_id"]
      .count()
      .sort_values(ascending=False)
)

print(supplier_relationship)

supplier_id
SUP0005    3044
SUP0006    3038
SUP0003    3033
SUP0007    3028
SUP0008    3010
SUP0004    2966
SUP0001    2944
SUP0002    2937
Name: po_id, dtype: int64


In [ ]:
# Purchase orders per branch
branch_relationship = (
    df.groupby("branch_id")["po_id"]
      .count()
      .sort_values(ascending=False)
)

print(branch_relationship)

branch_id
PUN001    4039
CHN001    4018
HYD001    4011
AHM001    4007
DEL001    4006
KOL001    3919
Name: po_id, dtype: int64


The relationship between suppliers and purchase orders is expected to be:

*   One Supplier → Many Purchase Orders


The relationship between branches and purchase orders is also expected to be:


*   One Branch → Many Purchase Orders


4.3 Relationship Map

+----------------------+
|   Supplier Master    |
|----------------------|
| PK: supplier_id      |
+----------+-----------+
           |
           | 1
           |
           | many
           v
+----------------------+
| Purchase Order Header|
|----------------------|
| PK: po_id            |
| FK: supplier_id      |
| FK: branch_id        |
| order_date           |
| expected_delivery    |
| received_date        |
| po_status            |
| total_cost           |
| GST                  |
| grand_total          |
+----------+-----------+
           |
           | many
           |
           | 1
           v
+----------------------+
|    Branch Master     |
|----------------------|
| PK: branch_id        |
+----------------------+





5. **Data** **Cleaning** / **Preparation**

5.1 Data Type Standardization

In [ ]:
# ID columns as strings
id_columns = [
    "po_id",
    "supplier_id",
    "branch_id"
]

for column in id_columns:
    df[column] = df[column].astype("string").str.strip()

# Date columns
date_columns = [
    "order_date",
    "expected_delivery_date",
    "received_date"
]

for column in date_columns:
    df[column] = pd.to_datetime(
        df[column],
        errors="coerce"
    )

# Numeric columns
financial_columns = [
    "total_cost",
    "total_gst_amount",
    "grand_total"
]

for column in financial_columns:
    df[column] = pd.to_numeric(
        df[column],
        errors="coerce"
    )

print(df.dtypes)

po_id                     string[python]
supplier_id               string[python]
branch_id                 string[python]
order_date                datetime64[ns]
expected_delivery_date    datetime64[ns]
received_date             datetime64[ns]
po_status                         object
total_cost                       float64
total_gst_amount                 float64
grand_total                      float64
calculated_grand_total           float64
dtype: object


The dataset is standardized so that:

* IDs are string values.
* Dates are datetime values.
* Financial fields are numeric.

5.2 Invalid Value Handling

In [ ]:
valid_statuses = ["Received", "Cancelled"]

invalid_status = ~df["po_status"].isin(valid_statuses)

print("Invalid status records:", invalid_status.sum())

Invalid status records: 0


No invalid status categories were found.

The missing received_date values correspond to cancelled purchase orders, so they should not automatically be replaced with another date.

5.3 Duplicate Handling

In [ ]:
before = len(df)

df = df.drop_duplicates()

after = len(df)

print("Rows before:", before)
print("Rows after:", after)
print("Rows removed:", before - after)

Rows before: 24000
Rows after: 24000
Rows removed: 0


No duplicate rows were removed.

5.4 Category Standardization

In [ ]:
# Remove unnecessary spaces
df["po_status"] = df["po_status"].str.strip()

# Standardize capitalization
df["po_status"] = df["po_status"].str.title()

# Check final categories
print(df["po_status"].value_counts())

po_status
Received     21630
Cancelled     2370
Name: count, dtype: int64


5.5 Final Validation

In [ ]:
print("Dataset shape:", df.shape)

print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

print("\nDuplicate PO IDs:")
print(df["po_id"].duplicated().sum())

print("\nStatuses:")
print(df["po_status"].value_counts())

print("\nNegative financial values:")

for column in financial_columns:
    print(
        column,
        (df[column] < 0).sum()
    )

print("\nDate consistency:")
print(
    "Order > Expected:",
    (df["order_date"] > df["expected_delivery_date"]).sum()
)

print(
    "Received < Order:",
    (
        df.loc[df["received_date"].notna(), "received_date"]
        <
        df.loc[df["received_date"].notna(), "order_date"]
    ).sum()
)

print("\nGrand total mismatches:")

grand_total_check = ~np.isclose(
    df["grand_total"],
    df["total_cost"] + df["total_gst_amount"],
    atol=0.02
)

print(grand_total_check.sum())

Dataset shape: (24000, 11)

Missing values:
po_id                        0
supplier_id                  0
branch_id                    0
order_date                   0
expected_delivery_date       0
received_date             2370
po_status                    0
total_cost                   0
total_gst_amount             0
grand_total                  0
calculated_grand_total       0
dtype: int64

Duplicate rows:
0

Duplicate PO IDs:
0

Statuses:
po_status
Received     21630
Cancelled     2370
Name: count, dtype: int64

Negative financial values:
total_cost 0
total_gst_amount 0
grand_total 0

Date consistency:
Order > Expected: 0
Received < Order: 0

Grand total mismatches:
0


The final dataset passes the major structural, financial, and date validation checks.

6. Documentation

6.1 Key Findings


The dataset contains 24,000 purchase-order records and 10 columns.

The main identifier is po_id, which contains 24,000 unique and non-null values. There are 8 suppliers and 6 branches represented in the dataset.

The purchase-order status contains two categories:

* Received — 21,630 records
* Cancelled — 2,370 records

The received_date field contains 2,370 missing values, corresponding to the cancelled purchase orders.

6.2 Data Quality Findings

The main data-quality findings are

Rows- 24,000

Columns - 10

Duplicate rows - 0

Duplicate po_id - 0

Missing po_id - 0

Missing received_date - 2,370

Invalid statuses - 0

Negative financial values - 0

Order date > expected date - 0

Received date < order date - 0

Grand-total calculation errors- 0

Late deliveries - 3,267

The 2,370 missing received dates are consistent with the 2,370 cancelled purchase orders, rather than being treated automatically as an error.


6.3 Cleaning Summary

The data-preparation process included:

1. Converting ID fields to string format.
2. Converting date fields to datetime format.
3. Converting financial 5fields to numeric format.
4. Removing unnecessary whitespace.
5. Standardizing purchase-order status values.
6. Checking duplicate records.
7. Validating primary-key uniqueness.
8. Checking financial calculations.
9. Validating date relationships.
10. Performing final data-quality checks.

No duplicate records required removal.


In [ ]:
df.to_csv("purchase_orders_header_cleaned.csv", index=False)

print("Cleaned CSV saved successfully!")

Cleaned CSV saved successfully!


In [ ]:
from google.colab import files

files.download("purchase_orders_header_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>